src/generators/r/httr2.ts

```diff
@@ -0,0 +1,268 @@
+import { Word, eq } from "../../shell/Word.js";
+import { parse, getFirst, COMMON_SUPPORTED_ARGS } from "../../parse.js";
+import type { Request, Warnings } from "../../parse.js";
+import { wordDecodeURIComponent, parseQueryString } from "../../Query.js";
+
+import { reprBacktick, reprStr, repr, toNumeric } from "./httr.js";
+
+export const supportedArgs = new Set([
+  ...COMMON_SUPPORTED_ARGS,
+  "form",
+  "form-string",
+  "insecure",
+  "no-insecure",
+  "oauth2-bearer",
+
+  "max-redirs",
+
+  "max-time",
+  "connect-timeout",
+
+  "retry",
+  "retry-max-time",
+]);
+
+type NamedArg = [Word | string, Word | string];
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+function getCookieList(request: Request, warnings: Warnings): Array<NamedArg> {
+  if (!request.cookies) {
+    return [];
+  }
+  const cookieList: Array<NamedArg> = [];
+  for (const [key, value] of request.cookies) {
+    try {
+      // httr2 percent-encodes cookie values
+      const decoded = wordDecodeURIComponent(value.replace(/\+/g, " "));
+      cookieList.push([key, decoded]);
+    } catch {
+      warnings.push([
+        "cookie-decoding error",
+        "Cookie could not be decoded\n\n" + value.toString(),
+      ]);
+    }
+  }
+
+  request.headers.delete("Cookie");
+  return cookieList;
+}
+
+function getQueryList(request: Request): Array<NamedArg> {
+  return request.urls[0].queryList || [];
+}
+
+function getHeaderList(request: Request): Array<NamedArg> {
+  const headerList: Array<NamedArg> = [];
+
+  if (request.headers.length) {
+    for (const [headerName, headerValue] of request.headers) {
+      if (headerValue !== null) {
+        headerList.push([headerName, headerValue]);
+      }
+    }
+  }
+  return headerList;
+}
+
+function getMultipartParams(request: Request): Array<NamedArg> | undefined {
+  if (!request.multipartUploads) {
+    return undefined;
+  }
+  const parts = request.multipartUploads.map((m): NamedArg => {
+    let content;
+    if ("contentFile" in m) {
+      content = `curl::form_file(${repr(m.contentFile)})`;
+    } else {
+      content = repr(m.content);
+    }
+    return [m.name, content];
+  });
+
+  return parts;
+}
+
+function addBodyStep(
+  steps: string[],
+  request: Request,
+  contentType: string | null | undefined,
+): Array<string> {
+  if (request.multipartUploads) {
+    const params = getMultipartParams(request);
+    return addCurlStep(steps, "req_body_multipart", [], params);
+  }
+
+  if (!request.data) {
+    return steps;
+  }
+
+  if (request.data.startsWith("@") && !request.isDataRaw) {
+    const filePath = request.data.slice(1);
+    steps = addCurlStep(steps, "req_body_file", [repr(filePath)]);
+  } else {
+    const [parsedQueryString] = parseQueryString(request.data);
+    const dataIsList = parsedQueryString && parsedQueryString.length;
+    if (dataIsList) {
+      steps = addCurlStep(steps, "req_body_form", [], parsedQueryString);
```

Comment:
Similarly, this not checking the content-type header is not ideal, since the generated code might end up sending a different content-type.